ena-submission/src/ena_deposition/ena_submission_helper.py

```diff
@@ -47,93 +57,82 @@
-    ena_submission_url: str
-    ena_reports_service_url: str
-
-
-def get_ena_config(
-    ena_submission_username_default: str,
-    ena_submission_password_default: str,
-    ena_submission_url_default: str,
-    ena_reports_service_url_default: str,
-) -> ENAConfig:
-    import dotenv
-    dotenv.load_dotenv()
-    ena_submission_username = os.getenv("ENA_USERNAME")
-    if not ena_submission_username:
-        ena_submission_username = ena_submission_username_default
-
-    ena_submission_password = os.getenv("ENA_PASSWORD")
-    if not ena_submission_password:
-        ena_submission_password = ena_submission_password_default
-
-    ena_submission_url = ena_submission_url_default
-    ena_reports_service_url = ena_reports_service_url_default
-
-    db_params = {
-        "ena_submission_username": ena_submission_username,
-        "ena_submission_password": ena_submission_password,
-        "ena_submission_url": ena_submission_url,
-        "ena_reports_service_url": ena_reports_service_url,
-    }
-
-    return ENAConfig(**db_params)
-
-
 @dataclass
 class CreationResult:
     errors: list[str]
     warnings: list[str]
-    result: dict[str, str] | None = None
+    result: dict[str, str | Sequence[str]] | None = None
 
 
 def recursive_defaultdict():
     return defaultdict(recursive_defaultdict)
 
 
-def dataclass_to_dict(dataclass_instance):
+class DataclassProtocol(Protocol):
+    __dataclass_fields__: ClassVar[dict[str, Field[Any]]]
+
+
+def assert_dataclass(obj: Any) -> None:
+    """
+    Asserts that the object is a dataclass instance.
+    Raises TypeError if not.
+    """
+    if not is_dataclass(obj):
+        msg = f"Expected a dataclass instance, got {type(obj).__name__}: {obj}."
+        raise TypeError(msg)
+
+
+def dataclass_to_dict(dataclass_instance: DataclassProtocol) -> dict[str, Any]:
     """
     Converts a dataclass instance to a dictionary, handling nested dataclasses.
     """
-    if not hasattr(dataclass_instance, "__dataclass_fields__"):
-        return dataclass_instance
-    result = {}
-    for field in dataclass_instance.__dataclass_fields__:
-        value = getattr(dataclass_instance, field)
-        is_xml_attribute = isinstance(value, XmlAttribute)
+    assert_dataclass(dataclass_instance)
+    result: dict[str, Any] = {}
+    for field_name in dataclass_instance.__dataclass_fields__:
+        value = getattr(dataclass_instance, field_name)
         if isinstance(value, XmlNone):
-            result[field.upper()] = None
+            result[field_name.upper()] = None
             continue
         if value is None:
             continue
         if isinstance(value, list):
-            result[field.upper()] = [dataclass_to_dict(item) for item in value]
-        elif is_xml_attribute:
-            attribute_field = "@" + field
+            res = []
+            for item in value:
+                assert_dataclass(item)
+                res.append(dataclass_to_dict(cast(DataclassProtocol, item)))
```

Comment:
why is this cast required?

Reply:
also managed to fix the other cast - this function is a bit of a mess, might make sense rewriting at some point. I think it dates back to the earliest experiments with ena submission :)